fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateExtractor.java

```diff
@@ -0,0 +1,193 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation.materialization;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Range;
+import com.google.common.collect.TreeRangeSet;
+import com.starrocks.analysis.BinaryType;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+
+import java.util.List;
+import java.util.Optional;
+
+public class PredicateExtractor extends ScalarOperatorVisitor<RangePredicate, PredicateExtractor.PredicateExtractorContext> {
+    private final List<ScalarOperator> columnEqualityPredicates = Lists.newArrayList();
+    private final List<ScalarOperator> residualPredicates = Lists.newArrayList();
+
+    public List<ScalarOperator> getColumnEqualityPredicates() {
+        return columnEqualityPredicates;
+    }
+
+    public List<ScalarOperator> getResidualPredicates() {
+        return residualPredicates;
+    }
+
+    public static class PredicateExtractorContext {
+        private boolean isAnd = true;
+
+        public boolean isAnd() {
+            return isAnd;
+        }
+
+        public PredicateExtractorContext setAnd(boolean and) {
+            isAnd = and;
+            return this;
+        }
+    }
+
+    @Override
+    public RangePredicate visit(ScalarOperator scalarOperator, PredicateExtractorContext context) {
+        return null;
+    }
+
+    @Override
+    public RangePredicate visitBinaryPredicate(
+            BinaryPredicateOperator predicate, PredicateExtractorContext context) {
+        ScalarOperator left = predicate.getChild(0);
+        ScalarOperator right = predicate.getChild(1);
+        if (left.isColumnRef() && right.isConstantRef()) {
```

Comment:
How about right is const re f and left is column ref?

Reply:
when enter this logic, the scalar operator has been normalized, so here could only be left is column and right is constant. I will add protection logic here.